### ISSUE TITLE
CA1508 false-positive in negated pattern-match

### ISSUE THREAD

**Issue by daiplusplus** [dotnet/roslyn-analyzers#5100]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508)

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.203](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

With this code...

```
public IReadOnlyList<ResizedImage> ResizeImage( ILoadedImage loadedImage, Rect cropRect, Boolean includeOnlyCropped, params Size[] sizes )
{
	if( loadedImage == null ) throw new ArgumentNullException(nameof(loadedImage));
	if( !(loadedImage is MagickLoadedImage magickLoadedImage) ) throw new ArgumentException( "Value is not a " + nameof(MagickLoadedImage) + ".", paramName: nameof(loadedImage));

	// ... etc
}
```

...CA1508 incorrectly reports this warning:

> CA1508 - 'loadedImage is MagickLoadedImage magickLoadedImage' is always 'true'. Remove or refactor the condition(s) to avoid dead code.

When I first saw the message I thought this was something to do with "sealed types" / closed-types, but then I remembered I'm still targeting .NET Core 3.1, not .NET 6+ - and also that:
* `ILoadedImage` is a public interface with multiple implementations besides `MagickLoadedImage`
* This method `ResizeImage` is a public method in a public type.
* Therefore, if `loadedImage` is not null it can still be a type other than `MagickLoadedImage`.

Also, the rule does not complain if I expand the statement to a multi-line `if` statement, as seen in the screenshot:

![image](https://user-images.githubusercontent.com/1693078/118261588-9ab3b880-b468-11eb-9341-16be2dc68fc8.png)

**Comment by Youssef1313** [comment]
> Also, the rule does not complain if I expand the statement to a multi-line if statement, as seen in the screenshot:

I doubt this is the case. It might be because of the usage of `== null` vs `is null`. Though the conclusion is the same, this indeed looks like a bug.

**Comment by mavasani** [comment]
Need to check against the latest preview package for drops from master: https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet6&package=Microsoft.CodeAnalysis.NetAnalyzers&version=6.0.0-preview6.21270.1&protocolType=NuGet&view=versions

**Comment by Truckerek** [comment]
Similar case and proof that it is not a dead conditional code:
`[TestMethod]
[ExpectedException(typeof(ArgumentException))]
public void DoTest()
{
	DoTest(string.Empty);
}

private static void DoTest(string value)
{
	if (value == null)
	{
		throw new ArgumentNullException(nameof(value));
	}
#pragma warning disable CA1508 // Avoid dead conditional code
	if (string.IsNullOrWhiteSpace(value))
#pragma warning restore CA1508 // Avoid dead conditional code
	{
		throw new ArgumentException(null, nameof(value));
	}
}`

**Comment by mavasani** [comment]
Confirmed that is fixed with #6805 and will be part of the next .NET 8 preview SDK.
